### PR TITLE
Create stronger type to expression compilation tracker internal policy

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Extracted from https://github.com/dotnet/roslyn/pull/72494.

Allows us to express succintly teh behavior we want in the compilation tracker around running SGs or generating skeleton references.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CreationPolicy.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        /// <summary>
+        /// Skeleton references should be created, and should be up to date with the project they are created for.
+        /// </summary>
+        Create,
+
+        /// <summary>
+        /// Skeleton references should only be created for a compilation if no existing skeleton exists for their
+        /// project from some point in the past.
+        /// </summary>
+        CreateIfAbsent,
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CreationPolicy.cs:39]: Note: this state will be used in teh followup PR.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs [hunk 2]
```diff
@@ -164,20 +167,20 @@ private sealed class FinalCompilationTrackerState : CompilationTrackerState
...
                 UnrootedSymbolSet unrootedSymbolSet)
-                : base(isFrozen, generatorInfo)
+                : base(creationPolicy, generatorInfo)
             {
                 Contract.ThrowIfNull(finalCompilationWithGeneratedDocuments);
 
                 // As a policy, all partial-state projects are said to have incomplete references, since the
                 // state has no guarantees.
                 this.CompilationWithoutGeneratedDocuments = compilationWithoutGeneratedDocuments;
-                HasSuccessfullyLoaded = hasSuccessfullyLoaded && !isFrozen;
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs:180]: >&& !isFrozen; [](http://example.com/codeflow?start=62&length=13)

Just wanted to make sure this was intentional

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs [hunk 3]
```diff
@@ -667,13 +674,18 @@ public ICompilationTracker FreezePartialState(CancellationToken cancellationToke
             {
                 var state = this.ReadState();
 
+                // We're freezing the solution for features where latency performance is parameter.  Do not run SGs or
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:677]: yes :D

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs:677]: will fix in followup pr.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs:180]: yes it was.  deciding to freeze things doesn't have any impact on our determination if things successfully loaded or not.

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski 	For review when you get back.

**Pull request merged**